File: asap-precompute-go/frame_identity.go
package precompute

import (
	"errors"
	"fmt"
	"reflect"
	"strconv"
	"sync"
)

// SummaryFrameIdentity is the plan-derived identity attached to one summary data point.
type SummaryFrameIdentity struct {
	IdentityVersion     uint32
	PlanID              uint64
	PlanVersion         uint64
	BackendCompat       string
	Materialization     uint64
	SeriesIdentity      string
	SchemaID            string
	ProducerID          string
	ProducerEpoch       string
	WindowStartUnixNano uint64
	WindowEndUnixNano   uint64
	Sequence            uint64
	Kind                string
	Encoding            StateEncoding
	CheckpointID        string
	BaseCheckpointID    string
}

type frameLineageKey struct {
	planID          uint64
	planVersion     uint64
	materialization uint64
	seriesIdentity  string
	producerID      string
	producerEpoch   string
}

type frameLineageState struct {
	sequence       uint64
	checkpointID   string
	checkpointAtMS uint64
}

// FrameSequencer owns sender ordering for all concrete series in one process.
type FrameSequencer struct {
	mu       sync.Mutex
	lineages map[frameLineageKey]frameLineageState
}

// FrameEmission describes one already-serialized envelope awaiting identity.
type FrameEmission struct {
	SeriesIdentity      string
	WindowStartUnixNano uint64
	WindowEndUnixNano   uint64
	NowUnixMS           uint64
	EmittedFull         bool
}

// NextBatchForEmission allocates a batch transactionally. If any envelope is
// invalid, no lineage sequence or checkpoint state is advanced.
func (s *FrameSequencer) NextBatchForEmission(plan CollectorPlan, rule TransmissionRule, producerEpoch string, emissions []FrameEmission) ([]SummaryFrameIdentity, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	cloned := make(map[frameLineageKey]frameLineageState, len(s.lineages))
	for key, state := range s.lineages {
		cloned[key] = state
	}
	candidate := FrameSequencer{lineages: cloned}
	frames := make([]SummaryFrameIdentity, 0, len(emissions))
	for _, emission := range emissions {
		frame, err := candidate.NextForEmission(
			plan, rule, producerEpoch, emission.SeriesIdentity,
			emission.WindowStartUnixNano, emission.WindowEndUnixNano,
			emission.NowUnixMS, emission.EmittedFull,
		)
		if err != nil {
			return nil, err
		}
		frames = append(frames, frame)
	}
	s.lineages = candidate.lineages
	return frames, nil
}

// Next returns a full checkpoint for the first frame in every lineage and
// periodic recovery checkpoints for delta rules.
func (s *FrameSequencer) Next(plan CollectorPlan, rule TransmissionRule, producerEpoch, seriesIdentity string, windowStart, windowEnd, nowUnixMS uint64) (SummaryFrameIdentity, error) {
	return s.next(plan, rule, producerEpoch, seriesIdentity, windowStart, windowEnd, nowUnixMS, nil)
}

// NextForEmission binds identity to the bytes the runtime actually emitted.
// A required full/checkpoint frame can never be mislabeled as a delta.
func (s *FrameSequencer) NextForEmission(plan CollectorPlan, rule TransmissionRule, producerEpoch, seriesIdentity string, windowStart, windowEnd, nowUnixMS uint64, emittedFull bool) (SummaryFrameIdentity, error) {
	return s.next(plan, rule, producerEpoch, seriesIdentity, windowStart, windowEnd, nowUnixMS, &emittedFull)
}

func (s *FrameSequencer) next(plan CollectorPlan, rule TransmissionRule, producerEpoch, seriesIdentity string, windowStart, windowEnd, nowUnixMS uint64, emittedFull *bool) (SummaryFrameIdentity, error) {
	if producerEpoch == "" || seriesIdentity == "" || windowStart >= windowEnd || rule.ProducerID != plan.CollectorID {
		return SummaryFrameIdentity{}, errors.New("invalid frame lineage")
	}
	matched := false
	for _, candidate := range plan.TransmissionRules {
		if reflect.DeepEqual(candidate, rule) {
			matched = true
			break
		}
	}
	if !matched {
		return SummaryFrameIdentity{}, errors.New("transmission rule does not belong to plan")
	}
	key := frameLineageKey{
		planID: plan.Envelope.PlanID, planVersion: plan.Envelope.PlanVersion,
		materialization: rule.Materialization, seriesIdentity: seriesIdentity,
		producerID: rule.ProducerID, producerEpoch: producerEpoch,
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.lineages == nil {
		s.lineages = make(map[frameLineageKey]frameLineageState)
	}
	state := s.lineages[key]
	state.sequence++
	checkpointDue := rule.FullCheckpointEveryMS != nil && nowUnixMS >= state.checkpointAtMS &&
		nowUnixMS-state.checkpointAtMS >= *rule.FullCheckpointEveryMS
	full := state.sequence == 1 || rule.Mode == TransmissionModeFull || checkpointDue
	if emittedFull != nil {
		if full && !*emittedFull {
			return SummaryFrameIdentity{}, errors.New("runtime emitted delta when a full checkpoint was required")
		}
		full = *emittedFull
	}
	kind := "delta"
	if full {
		kind = "full"
		state.checkpointID = fmt.Sprintf("%s:%s:%d:%d", rule.ProducerID, producerEpoch, windowStart, state.sequence)
		state.checkpointAtMS = nowUnixMS
	}
	s.lineages[key] = state
	frame := SummaryFrameIdentity{
		IdentityVersion: 1, PlanID: plan.Envelope.PlanID, PlanVersion: plan.Envelope.PlanVersion,
		BackendCompat: plan.Envelope.BackendCompat, Materialization: rule.Materialization,
		SeriesIdentity: seriesIdentity, SchemaID: rule.SchemaID, ProducerID: rule.ProducerID,
		ProducerEpoch: producerEpoch, WindowStartUnixNano: windowStart, WindowEndUnixNano: windowEnd,
		Sequence: state.sequence, Kind: kind, Encoding: rule.Encoding,
	}
	if full {
		frame.CheckpointID = state.checkpointID
	} else {
		frame.BaseCheckpointID = state.checkpointID
	}
	return frame, nil
}

// OTLPAttributes returns the exact reserved attribute names consumed by ASAPQuery-backend.
func (f SummaryFrameIdentity) OTLPAttributes() map[string]string {
	attrs := map[string]string{
		"asap.frame.identity_version": strconv.FormatUint(uint64(f.IdentityVersion), 10),
		"asap.frame.plan_id":          strconv.FormatUint(f.PlanID, 10),
		"asap.frame.plan_version":     strconv.FormatUint(f.PlanVersion, 10),
		"asap.frame.backend_compat":   f.BackendCompat,
		"asap.frame.materialization":  strconv.FormatUint(f.Materialization, 10),
		"asap.frame.series_identity":  f.SeriesIdentity,
		"asap.frame.schema_id":        f.SchemaID,
		"asap.frame.producer_id":      f.ProducerID,
		"asap.frame.producer_epoch":   f.ProducerEpoch,
		"asap.frame.sequence":         strconv.FormatUint(f.Sequence, 10),
		"asap.frame.kind":             f.Kind,
		"asap.frame.encoding":         string(f.Encoding),
	}
	if f.CheckpointID != "" {
		attrs["asap.frame.checkpoint_id"] = f.CheckpointID
	}
	if f.BaseCheckpointID != "" {
		attrs["asap.frame.base_checkpoint_id"] = f.BaseCheckpointID
	}
	return attrs
}

File: asap-precompute-rs/src/frame_identity.rs
//! Plan-derived summary-frame identity and sequence generation.

use std::collections::{BTreeMap, HashMap};

use serde::{Deserialize, Serialize};
use thiserror::Error;

use crate::collector_plan::{CollectorPlan, StateEncoding, TransmissionMode, TransmissionRule};

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[allow(missing_docs)]
pub enum SummaryFrameKind {
    Full,
    Delta,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
#[allow(missing_docs)]
pub struct SummaryFrameIdentity {
    pub identity_version: u32,
    pub plan_id: u64,
    pub plan_version: u64,
    pub backend_compat: String,
    pub materialization: u64,
    pub series_identity: String,
    pub schema_id: String,
    pub producer_id: String,
    pub producer_epoch: String,
    pub window_start_unix_nano: u64,
    pub window_end_unix_nano: u64,
    pub sequence: u64,
    pub kind: SummaryFrameKind,
    pub encoding: StateEncoding,
    pub checkpoint_id: Option<String>,
    pub base_checkpoint_id: Option<String>,
}

impl SummaryFrameIdentity {
    /// Encode the reserved modified-OTLP attributes consumed by the
    /// ASAPQuery backend. Window bounds remain the data-point timestamps.
    pub fn otlp_attributes(&self) -> BTreeMap<String, String> {
        let mut attributes = BTreeMap::from([
            (
                "asap.frame.identity_version".into(),
                self.identity_version.to_string(),
            ),
            ("asap.frame.plan_id".into(), self.plan_id.to_string()),
            (
                "asap.frame.plan_version".into(),
                self.plan_version.to_string(),
            ),
            (
                "asap.frame.backend_compat".into(),
                self.backend_compat.clone(),
            ),
            (
                "asap.frame.materialization".into(),
                self.materialization.to_string(),
            ),
            (
                "asap.frame.series_identity".into(),
                self.series_identity.clone(),
            ),
            ("asap.frame.schema_id".into(), self.schema_id.clone()),
            ("asap.frame.producer_id".into(), self.producer_id.clone()),
            (
                "asap.frame.producer_epoch".into(),
                self.producer_epoch.clone(),
            ),
            ("asap.frame.sequence".into(), self.sequence.to_string()),
            (
                "asap.frame.kind".into(),
                match self.kind {
                    SummaryFrameKind::Full => "full",
                    SummaryFrameKind::Delta => "delta",
                }
                .into(),
            ),
            (
                "asap.frame.encoding".into(),
                match self.encoding {
                    StateEncoding::SketchlibProtobufV1 => "sketchlib_protobuf_v1",
                    StateEncoding::SketchCoreMsgpackV1 => "sketch_core_msgpack_v1",
                    StateEncoding::ExactAccumulatorV1 => "exact_accumulator_v1",
                }
                .into(),
            ),
        ]);
        if let Some(checkpoint_id) = &self.checkpoint_id {
            attributes.insert("asap.frame.checkpoint_id".into(), checkpoint_id.clone());
        }
        if let Some(base_checkpoint_id) = &self.base_checkpoint_id {
            attributes.insert(
                "asap.frame.base_checkpoint_id".into(),
                base_checkpoint_id.clone(),
            );
        }
        attributes
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
struct LineageKey {
    materialization: u64,
    series_identity: String,
    producer_id: String,
    producer_epoch: String,
    window_start_unix_nano: u64,
    window_end_unix_nano: u64,
}

#[derive(Debug, Clone)]
struct LineageState {
    sequence: u64,
    checkpoint_id: String,
    checkpoint_at_unix_ms: u64,
}

#[derive(Debug, Error, PartialEq, Eq)]
#[allow(missing_docs)]
pub enum FrameIdentityError {
    #[error("transmission rule does not belong to this plan/collector")]
    RuleMismatch,
    #[error("frame window is invalid")]
    InvalidWindow,
}

/// One sequencer belongs to one running collector process. A new process uses
/// a new producer epoch, so backend lineage cannot confuse pre/post-restart
/// sequences.
#[derive(Debug, Default)]
pub struct FrameSequencer {
    lineages: HashMap<LineageKey, LineageState>,
}

impl FrameSequencer {
    /// Produce the next full/delta identity for one plan-bound lineage.
    pub fn next(
        &mut self,
        plan: &CollectorPlan,
        rule: &TransmissionRule,
        producer_epoch: &str,
        series_identity: &str,
        window_start_unix_nano: u64,
        window_end_unix_nano: u64,
        now_unix_ms: u64,
    ) -> Result<SummaryFrameIdentity, FrameIdentityError> {
        if rule.producer_id != plan.collector_id
            || !plan
                .transmission_rules
                .iter()
                .any(|candidate| candidate == rule)
            || producer_epoch.is_empty()
            || series_identity.is_empty()
        {
            return Err(FrameIdentityError::RuleMismatch);
        }
        if window_start_unix_nano >= window_end_unix_nano {
            return Err(FrameIdentityError::InvalidWindow);
        }
        let key = LineageKey {
            materialization: rule.materialization,
            series_identity: series_identity.into(),
            producer_id: rule.producer_id.clone(),
            producer_epoch: producer_epoch.into(),
            window_start_unix_nano,
            window_end_unix_nano,
        };
        let state = self.lineages.entry(key).or_insert_with(|| LineageState {
            sequence: 0,
            checkpoint_id: String::new(),
            checkpoint_at_unix_ms: 0,
        });
        state.sequence = state.sequence.saturating_add(1);
        let checkpoint_due = state.sequence == 1
            || rule.mode == TransmissionMode::Full
            || rule.full_checkpoint_every_ms.is_some_and(|cadence| {
                now_unix_ms.saturating_sub(state.checkpoint_at_unix_ms) >= cadence
            });
        let kind = if checkpoint_due {
            SummaryFrameKind::Full
        } else {
            SummaryFrameKind::Delta
        };
        if kind == SummaryFrameKind::Full {
            state.checkpoint_id = format!(
                "{}:{}:{}:{}",
                rule.producer_id, producer_epoch, window_start_unix_nano, state.sequence
            );
            state.checkpoint_at_unix_ms = now_unix_ms;
        }
        Ok(SummaryFrameIdentity {
            identity_version: 1,
            plan_id: plan.envelope.plan_id,
            plan_version: plan.envelope.plan_version,
            backend_compat: plan.envelope.backend_compat.clone(),
            materialization: rule.materialization,
            series_identity: series_identity.into(),
            schema_id: rule.schema_id.clone(),
            producer_id: rule.producer_id.clone(),
            producer_epoch: producer_epoch.into(),
            window_start_unix_nano,
            window_end_unix_nano,
            sequence: state.sequence,
            kind,
            encoding: rule.encoding,
            checkpoint_id: (kind == SummaryFrameKind::Full).then(|| state.checkpoint_id.clone()),
            base_checkpoint_id: (kind == SummaryFrameKind::Delta)
                .then(|| state.checkpoint_id.clone()),
        })
    }
}
